File: n_by_n_tic_tac_toe/nbyn.py
matrix = [["x","o",""],
          ["x","o",""],
          ["x","","x"]]

# helper method, this the main method to see the check for n by n

"""

Since this is a symmetrical matrix it is easy to check rows and columns

It doesn't matter what the size of the matrix is since we just have to check the length of the row
and compare the 2.

"""

def check_x_and_y(board):
    x = "".join(["x" for i in range(len(board))])
    y = "".join(["o" for i in range(len(board))])
    return (x,y)

# helper method for diagonal check
def diagonal_condition_check(result,check):
    if "".join(result) == check[0]:
        return "x"
    elif "".join(result) == check[1]:
        return "o"

"""
for checking rows
"""
def row_check(board):
    check = check_x_and_y(board)

    for row in board:
        joined_row = "".join(row)
        if joined_row == check[0]:
            return "x"
        elif joined_row == check[1]:
            return "o"

    return None

"""
for checking columns
"""
def vertical_check(board):
    transposed_matrix = zip(*board)
    return row_check(transposed_matrix)

"""
for checking straight diagonal
"""
def diagonal_check_1(board):
    check = check_x_and_y(board)
    result = []
    i = 0

    for row in board:
        result.append(row[i])
        i += 1

    return diagonal_condition_check(result, check)

"""
for checking reverse diagonal
"""
def diagonal_check_2(board):
    check = check_x_and_y(board)
    result = []
    i = len(board)-1

    for row in board:
        result.append(row[i])
        i -= 1

    return diagonal_condition_check(result, check)

def main(board):
    if row_check(board):
        print row_check(board)
    elif vertical_check(board):
        print vertical_check(board)
    elif diagonal_check_1(board):
        print diagonal_check_1(board)
    elif diagonal_check_2(matrix):
        print diagonal_check_2(board)
    else:
        print "draw"

main(matrix)

File: n_by_n_tic_tac_toe/nbyn.go
package main

import "fmt"
import "strings"

func transpose(board [][]string) [][]string {
  transposed_matrix := make([][]string, len(board))

  for row := 0; row < len(board); row++ {
    transposed_matrix[row] = make([]string, len(board))
  }

  for i := 0; i < 3; i++ {
    for j := 0; j < 3; j++ {
      transposed_matrix[i][j] = board[j][i]
    }
  }

  return transposed_matrix
}

func get_joined_strings(board [][]string, option string) string {
  s := make([]string, 0)
  for i := 0; i < len(board); i++ {
    s  = append(s, option)
  }

  s_joined := strings.Join(s[:],"")

  return s_joined
}

func check_x_and_y(board [][]string) []string {
  x := get_joined_strings(board, "x")
  y := get_joined_strings(board, "y")
  results := make([]string, 0)
  results = append(results, x)
  results = append(results, y)

  return results
}

func row_check(board [][]string) string {
  check := check_x_and_y(board)

  for _, element := range board {
    joined_element := strings.Join(element[:],"")

    if joined_element == check[0] {
      return "x"
    } else if joined_element == check[1] {
      return "o"
    }
  }

  return "no checks"
}

func vertical_check(board [][]string) string {
  transposed_matrix := transpose(board)
  return row_check(transposed_matrix)
}

func main() {
  board := [][]string{ {"x", "o", "x"},
                       {"x", "x", "x"},
                       {"x", "x", "o"}, }

  fmt.Println(vertical_check(board))
}
